README.md
# Sphinx with Tex Live-2018 with Python 3.6.x

This image contains [Sphinx](http://www.sphinx-doc.org/en/master/index.html), [Python](https://www.python.org/) 3.6.x and [Tex Live](http://www.tug.org/texlive/) to help building PDF from [reStructedText](http://www.sphinx-doc.org/en/master/usage/restructuredtext/basics.html).

## Requirements

- [docker](https://docker.com)

This image should work with:

- Windows 10
- Windows Server 2016
- Linux (see How to get it section)

Please check *_Install Docker for Windows_* section for how to get Docker running.

## Quickstart

To build your documents:

```
docker run --rm -it -v C:/path/to/reStructedText:C:/docs hieunghiemba/texlive-2018-windows
```

All sub-directories in the source which contains the appropriate contents will be built.
You can find PDF file _'build/latex'_ of each built directory.

To have more details information, please check *_How to compile documents_* section.

## Documentation

- [Install Docker for Windows](https://docs.docker.com/docker-for-windows/install/)
- [Build from source](https://github.com/hieunba/get-sphinx/wiki/Build-from-source)
- [How to compile documents](https://github.com/hieunba/get-sphinx/wiki/How-to-compile)

## Note

The Windows image was built natively by Windows platform then it's quite big but you do not have to touch your system like installing libraries, or dependencies, etc.


CHANGELOG.md
# get-sphinx CHANGELOG

This file is used to list changes made by each version of get sphinx.

## 0.7.1 (2019-01-17)

  - Use gcim to deprecate Get-WmiObject command with Powershell 6 or later (following latest Docker image from Microsoft)
  - Update PATH to make.exe with new installation from Chocolatey (compatible with old PATH)

## 0.7.0 (2018-09-24)

  - Build filenames with spaces from hats (^)

## 0.6.2 (2018-09-20)

  - Clean up old artifacts before building

## 0.6.1 (2018-08-31)

  - Support custom source while building texlive 2017 for Windows

   To set Tex Live source, please add TEXLIVE_SOURCE in build-arg while issuing ` docker build `.

## 0.6.0 (2018-08-31)

  - Dockerfile for texlive 2018 for Windows
  - Collect artifacts in output in doc source

## 0.5.0 (2018-08-10)

  - Add Pandoc converter
  - Update build sequence to generate docx as well

## 0.4.0 (2018-07-25)

  - Build single HTTML by default

## 0.3.1 (2018-07-22)

  - Add Batch Build regardless of the depth levels of directories from input source

## 0.3.0 (2018-07-11)

  - Add Batch Build with the depth level of 1

## 0.2.2 (2018-07-05)

  - Add Cloud Sphinx Theme v1.9.4

## 0.2.1 (2018-07-05)

  - Add MinGW to image to use make

## 0.2.0 (2018-07-04)

  - Add image which includes Python 3.6.5, Sphinx and Tex Live 2017

## 0.1.0 (Unkown)

  - Powershell script to install Python 3.6.5, Sphinx and MikTex


test/smoke/default/default_test.rb
unless os.linux?
  # Tests check with py laucher for version 3 only
  describe command('py -3 --version') do
    its('exit_status') { should eq 0 }
    its('stdout') { should match (/Python 3.[5-9].[0-9]/) }
  end
  describe command('py -3 -m easy_install --version') do
    its('exit_status') { should eq 0 }
    its('stdout') { should match (/setuptools ([0-9][0-9]).[0-9].[0-9]/) }
  end
  describe command('sphinx-build --version') do
    its('exit_status') { should eq 0 }
    its('stdout') { should match (/sphinx-build 1.[0-9].[0-9]/) }
  end
  describe command('sphinx-quickstart --version') do
    its('exit_status') { should eq 0 }
    its('stdout') { should match (/sphinx-quickstart 1.[0-9].[0-9]/) }
  end
end


scripts/install.ps1
<#
  .SYNOPSIS
  Install Sphinx + MiKTeK

  .DESCRIPTION
  Install Sphinx based on Python 3.x and add MiKTeX to support LaTex.

  .EXAMPLE
  install.ps1

  .EXAMPLE
  install.ps1 -pythonVersion '3.6.5'

#>
param(
  [string]$pythonVersion = '3.6.5'
)


######################################
# Configuration parameters           #
######################################

$arch = $env:PROCESSOR_ARCHITECTURE.ToLower()

function Install-MikTex {
    <#
    .SYNOPSIS
    Install MikTex

    .DESCRIPTION
    Install MikTex 2.9.6643-x64. Currently supports installing from pre-built binary file.

    .EXAMPLE
    Install-MikTex

    #>
    param(
        [Boolean]
        $Source = $false,
        [String]
        $Repository = "$env:TMP\miktex"
    )

    $miktex_reg_path = 'HKLM:\SOFTWARE\Microsoft\Windows\CurrentVersion\Uninstall\MikTex 2.9'
    if (Test-RegistryKeyValue -Path $miktex_reg_path -Name 'MikTex 2.9') {
        Write-Host "MikTex 2.9 installed." -ForegroundColor Magenta
        return $true
    }

    if (-Not $Source) {
        $miktex_amd64_setup_url = 'https://www.dropbox.com/s/vj2g3c4hih8gabb/basic-miktex-2.9.6643-x64.exe?dl=1'
        $miktex_amd64_setup_checksum = '792983b8945ddafc5285cb9942d9d88550ef3af0f8d4add9acd057233ff84584'
        $miktex_download_path = "$env:TMP\basic-miktex-setup.exe"
        if (-Not (Test-Path $miktex_download_path))
        {
            Write-Host "Downloading MiKTeX installation binary..." -ForegroundColor Magenta
            (New-Object System.Net.WebClient).DownloadFile($miktex_amd64_setup_url, $miktex_download_path)
        }
        else
        {
            $fileHash = Get-FileHash $miktex_download_path -Algorithm SHA256
            if ($fileHash -eq $miktex_amd64_setup_checksum) {
                Write-Host "MiKTeX installation binary staged. Skip downloading." -ForegroundColor Yellow
            }
            else
            {
                Write-Host "Checksum mismatched. Redownloading MiKTeX installation binary... " -ForegroundColor Magenta
                (New-Object System.Net.WebClient).DownloadFile($miktex_amd64_setup_url, $miktex_download_path)
            }
        }
        Write-Host "Installng MikTex..." -ForegroundColor Magenta
        Start-Process $miktex_download_path -ArgumentList @('--unattended', '--auto-install=yes') -Wait
    }
    else
    {

        $deploy_setup_url = 'https://www.dropbox.com/s/aavayoshtp2kzp0/miktexsetup.amd64.exe?dl=1'
        $deploy_setup_checksum = 'da42decf97617682c5a02b84260c46bb80b9ff9cd26e365ab0c0d9619c38387c'
        $deploy_setup_path = "$env:TMP\miktex_deploy.exe"

        if ($deploy_setup_path) {
            if (-Not ((Get-FileHash $deploy_setup_path -Algorithm SHA256) -eq $deploy_setup_checksum)) {
                Write-Host "Downloading MikTex deploy utility..." -ForegroundColor Magenta
                (New-Object System.Net.WebClient).DownloadFile($deploy_setup_url, $deploy_setup_path)
            }
            else
            {
                Write-Host "MikTek deploy utility staged. Skipped downloading."
            }
        }

        Write-Host "Preparing MikTek local repository at $Repository" -ForeGroundColor Magenta
        Start-Process $deploy_setup_path -ArgumentList @('--verbose', "--local-package-repository=$Repository --package-set=complete download") -Wait
        Write-Host "Installing MikTek from local repository..." -ForeGroundColor Magenta
        Start-Process $deploy_setup_path -ArgumentList @('--quiet', "--local-package-repository=$Repository --package-set=basic install") -Wait

    }
}

function Test-RegistryKeyValue
{
    <#
    .SYNOPSIS
    Tests if a registry value exists or a product has installed which has a registry value exists in Uninstall.

    .DESCRIPTION
    The usual ways for checking if a registry value exists don't handle when a value simply has an empty or null value.  This function actually checks if a key has a value with a given name.

    .EXAMPLE
    Test-RegistryKeyValue -Path 'hklm:\Software\Carbon\Test' -Name 'Title'

    Returns `True` if `hklm:\Software\Carbon\Test` contains a value named 'Title'.  `False` otherwise.

    .EXAMPLE
    Test-RegistryKeyValue -Path `hklm:\Software\Microsoft\Windows\CurrentVersion\Uninstall\Milk` -Name `Milk`

    Returns `True` if `hklm:\Software\Microsoft\Windows\CurrentVersion\Uninstall\Milk` or
    if it contains `DisplayName` has a value of 'Milk'. `False` otherwise.

    .DERIVE
    This was referred to the work of Aaron Jensen on stackoverflow, or at
    https://stackoverflow.com/questions/5648931/test-if-registry-value-exists#5652674
    #>
    [CmdletBinding()]
    param(
        [Parameter(Mandatory=$true)]
        [string]
        # The path to the registry key where the value should be set.  Will be created if it doesn't exist.
        $Path,

        [Parameter(Mandatory=$true)]
        [string]
        # The name of the value being set.
        $Name
    )

    if( -not (Test-Path -Path $Path -PathType Container) )
    {
        return $false
    }

    $properties = Get-ItemProperty -Path $Path
    if( -not $properties )
    {
        return $false
    }

    if ($properties.DisplayName -eq $Name)
    {
        return $true
    }

    $member = Get-Member -InputObject $properties -Name $Name
    if( $member )
    {
        return $true
    }
    else
    {
        return $false
    }

}

function Find-Python {
  # Check if Python 3.x installed or not

  $currentVersion = (gwmi -Class Win32_Product | Where { $_.Name -match 'Python.*Executables' }).Version

  if ($currentVersion -match $pythonVersion.SubString(0,1)) {
    return $true;
  } elseif ($currentVersion -match $pythonVersion) {
    return $true;
  } else {
    return $false;
  }
}

function Install-Python {

  $version = $pythonVersion

  $python_exe_path 	  = "$env:PROGRAMFILES\Python36\python.exe"
  $python_download_url  = "https://www.python.org/ftp/python/$version/python-$version-$arch.exe"
  $python_download_path = "$env:TMP\python-$version-$arch.exe"

  if (-Not (Test-Path $python_download_path)) {
    (New-Object System.Net.WebClient).DownloadFile($python_download_url, $python_download_path)
  }

  if (-Not $(Find-Python)) {
    Write-Host "Python was missing. Install python $version..." -ForegroundColor Magenta
    Start-Process $python_download_path -ArgumentList @('/quiet', "InstallAllUsers=1 PrependPath=1 Include_test=0") -Wait
  }
}

function Install-Sphinx {
  $easy_install_exe_path = "$env:PROGRAMFILES\Python36\Scripts\easy_install.exe"
  Start-Process py -ArgumentList @('-3', '-m pip', 'install sphinx') -Wait
}

# Start
# Need to work with TLSv1.2 also
[System.Net.ServicePointManager]::SecurityProtocol = [System.Net.SecurityProtocolType]::Tls11 -bor [System.Net.SecurityProtocolType]::Tls12;
Install-Python
Install-Sphinx
Install-MikTex


bootstrap.ps1


 ######################################
 # Configuration parameters           #
 ######################################

 ######################################
 # Notes                              #
 ######################################
 # ez_setup.py used to install easy_install which is hosted at https://bootstrap.pypa.io/ez_setup.py
 # Due to the implementation of TLS v1.2, then need to upload to Dropbox
 $arch = $env:PROCESSOR_ARCHITECTURE.ToLower()
 $py27_dir = "C:\Python27"
 $py27_exe_path = "$py27_dir\python.exe"
 $easy_install_exe_path = "$py27_dir\Scripts\easy_install.exe"
 $default_download_path = $env:TMP

function Install-MikTex {
    <#
    .SYNOPSIS
    Install MikTex

    .DESCRIPTION
    Install MikTex 2.9.6643-x64. Currently supports installing from pre-built binary file.

    .EXAMPLE
    Install-MikTex

    #>
    [CmdletBinding()]
    param(
        [Parameter(Mandatory=$false)]
        [Boolean]
        $Source = $false,
        [Parameter(Mandatory=$false)]
        [String]
        $Repository = "$env:TMP\miktex"
    )
    $miktex_reg_path = 'HKLM:\SOFTWARE\Microsoft\Windows\CurrentVersion\Uninstall\MikTex 2.9'
    if (Test-RegistryKeyValue -Path $miktex_reg_path -Name 'MikTex 2.9') {
        Write-Host "MikTex 2.9 installed." -ForegroundColor Magenta
        return $true
    }

    if (-Not $Source) {
        $miktex_amd64_setup_url = 'https://www.dropbox.com/s/vj2g3c4hih8gabb/basic-miktex-2.9.6643-x64.exe?dl=1'
        $miktex_amd64_setup_checksum = '792983b8945ddafc5285cb9942d9d88550ef3af0f8d4add9acd057233ff84584'
        $miktex_download_path = "$env:TMP\basic-miktex-setup.exe" 

        if ($arch -eq 'amd64') {
            $fileHash = Get-FileHash $miktex_download_path -Algorithm SHA256
            if ($fileHash -eq $miktex_amd64_setup_checksum) {
                Write-Host "MixTek installation binary staged. Skip downloading." -ForegroundColor Yellow
            } else 
            {
                (New-Object System.Net.WebClient).DownloadFile($miktex_amd64_setup_url, $miktex_download_path)
            }
            Write-Host "Installng MikTex.." -ForegroundColor Magenta
            Start-Process $miktex_download_path -ArgumentList @('--unattended', '--auto-install=yes') -Wait
        }
        else 
        {
            Write-Host "MikTex does not support $env:PROCESSOR_ARCHITECTURE platform." -ForegroundColor Yellow
        }
    }
    else 
    {
  
        $deploy_setup_url = 'https://www.dropbox.com/s/aavayoshtp2kzp0/miktexsetup.amd64.exe?dl=1'
        $deploy_setup_checksum = 'da42decf97617682c5a02b84260c46bb80b9ff9cd26e365ab0c0d9619c38387c'
        $deploy_setup_path = "$env:TMP\miktex_deploy.exe"
 
        if ($deploy_setup_path) {
            if (-Not ((Get-FileHash $deploy_setup_path -Algorithm SHA256) -eq $deploy_setup_checksum)) {
                Write-Host "Downloading MikTex deploy utility.." -ForegroundColor Magenta
                (New-Object System.Net.WebClient).DownloadFile($deploy_setup_url, $deploy_setup_path)
            }
            else
            {
                Write-Host "MikTek deploy utility staged. Skipped downloading."
            }
        }    
      
        Write-Host "Preparing MikTek local repository at $Repository" -ForeGroundColor Magenta
        Start-Process $deploy_setup_path -ArgumentList @('--verbose', "--local-package-repository=$Repository --package-set=complete download") -Wait
        Write-Host "Installing MikTek from local repository..." -ForeGroundColor Magenta
        Start-Process $deploy_setup_path -ArgumentList @('--quiet', "--local-package-repository=$Repository --package-set=basic install") -Wait
  
    }
}

function Test-RegistryKeyValue
{
    <#
    .SYNOPSIS
    Tests if a registry value exists or a product has installed which has a registry value exists in Uninstall.

    .DESCRIPTION
    The usual ways for checking if a registry value exists don't handle when a value simply has an empty or null value.  This function actually checks if a key has a value with a given name.

    .EXAMPLE
    Test-RegistryKeyValue -Path 'hklm:\Software\Carbon\Test' -Name 'Title'

    Returns `True` if `hklm:\Software\Carbon\Test` contains a value named 'Title'.  `False` otherwise.

    Test-RegistryKeyValue -Path `hklm:\Software\Microsoft\Windows\CurrentVersion\Uninstall\Milk` -Name `Milk`

    Returns `True` if `hklm:\Software\Microsoft\Windows\CurrentVersion\Uninstall\Milk` or 
    if it contains `DisplayName` has a value of 'Milk'. `False` otherwise.

    .DERIVE
    This was referred to the work of Aaron Jensen on stackoverflow, or at 
    https://stackoverflow.com/questions/5648931/test-if-registry-value-exists#5652674
    #>
    [CmdletBinding()]
    param(
        [Parameter(Mandatory=$true)]
        [string]
        # The path to the registry key where the value should be set.  Will be created if it doesn't exist.
        $Path,

        [Parameter(Mandatory=$true)]
        [string]
        # The name of the value being set.
        $Name
    )

    if( -not (Test-Path -Path $Path -PathType Container) )
    {
        return $false
    }

    $properties = Get-ItemProperty -Path $Path 
    if( -not $properties )
    {
        return $false
    }

    if ($properties.DisplayName -eq $Name)
    {
        return $true
    }

    $member = Get-Member -InputObject $properties -Name $Name
    if( $member )
    {
        return $true
    }
    else
    {
        return $false
    }

}

function Install-Python {
  <#
  .SYNOPSIS
  Install Python for Windows and install easy_install

  .DESCRIPTION
  By default, installing Pyhon 2.7.14 if no version was specified

  .EXAMPLE
  Install-Python

  #>

  [CmdletBinding()]
  param(
      [Parameter(Mandatory=$false)]
      [string]
      $version = "2.7.14"
  )

  $python_exe_path      = "C:\Python27\python.exe"
  $python_download_url  = "https://www.python.org/ftp/python/$version/python-$version.$arch.msi"
  $python_download_path = "$env:TMP\python-$version.$arch.msi"

  if (-Not (Test-Path $python_download_path)) {
    (New-Object System.Net.WebClient).DownloadFile($python_download_url, $python_download_path)
  }
 
  Get-Python-Setup-Files

  foreach ($member in (Get-WmiObject -Class Win32_Product)) { if ($member.Name -like '*Python*') { $python_installed = $true } }

  if (-Not $python_installed) {
    Start-Process msiexec.exe -ArgumentLIst @('/q', "/i $python_download_path") -Wait
  }
  if (-Not (Test-Path $easy_install_exe_path)) {
    Start-Process python.exe -ArgumentList @("$Location\ez_setup.py") -Wait
  }
}

function Install-Sphinx {
  if (-Not (Test-Path $py27_exe_path)) {
    Write-Host "Python was missing. Installing it.." -ForegroundColor Magenta
    Install-Python
  }
  # Environment is not populated at this session
  if (Test-Path $easy_install_exe_path) {
    Start-Process $easy_install_exe_path -ArgumentList @('sphinx') -Wait
  } 
  else 
  {
    Write-Host "easy_install was not found. Please install it first." -ForegroundColor Red
  }
}

function Get-Python-Setup-Files {
    <#
    .SYNOPSIS
    Get Python setup files

    .DESCRIPTION
    Python setup files include ez_setup.py and get-pip.py. They are fetched from Internet and stores to TMP dir

    .EXAMPLE
    Get-Python-Setup-Files
    #>
    $easy_setup_url = "https://www.dropbox.com/s/14vztnkqf4ayhp3/ez_setup.py?dl=1"
    $get_pip_url = "https://raw.github.com/pypa/pip/master/contrib/get-pip.py"

    (New-Object System.Net.WebClient).DownloadFile($easy_setup_url, "$default_download_path\ez_setup.py")
    (New-Object System.Net.WebClient).DownloadFile($get_pip_url, "$default_download_path\get-pip.py")
    
    # If IE engine is available, can use Invoke-WebRequest to fetch file
    # but to be safe, using .Net framework one.
    #    Invoke-WebRequest -Uri $ez_setup_url -OutFile "distribute_setup.py"
}

# Prepare
if (-Not ([Environment]::GetEnvironmentVariable("Path") | Select-String Python27)) {
  [Environment]::SetEnvironmentVariable("Path", "$env:Path;C:\Python27;C:\Python27\Scripts\", "Machine")
}


# Start
Install-Python
Install-Sphinx
Install-MikTex